### PR TITLE
Job Model and Module

### PR DESCRIPTION
**Pull request by LokiFrostGiant**
## Tickets
#30 #53 

## Summary Of Changes
- Add Job module and model to backend. Add Deadline model for Jobs

## Notes for Reviewers
- Please look for comments asking questions in the change comparison
- Input on Job and Deadline models *critical*
- Any input on query/business/view very important

### CODE REVIEW

**Comment by jaredwebber** [comment]
I added the main structure of the view layer & placeholder business & query functions.
Breakdown is:
- `get_minimum_jobs`:
  - Param: `user_id`
  - Return: Array of `{id, position, company}` for each job the User has
- `get_job_by_id`:
  - Params: `user_id`, `job_id`
  - Return: Complete Job object associated with that `job_id` (`user_id` is for access permission validation)
- `create_job`:
  - Param: Job object constructed on frontend
  - Return: updated Job object (job `id` field will be updated on insertion into db)
- `update_job`:
  - Params: Job object
  - Return: Nothing

May be some tricky elements working with Job `Deadline` array -> query layer may need helper functions related to creating/updating/deleting deadlines. Could instead switch to JSONField, as that may be simpler to work with - just stringify the Deadline array passed from the frontend.

**Comment by LokiFrostGiant** [comment]
For `get_minimum_jobs` I am using `values` rather than `values_list`, favouring dictionaries over tuples.

File: backend/job/query.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+from django.db.models.query import QuerySet
+from .models import *
+
+
+def create_job(payload: dict) -> Job:
+    # need to return Job.objects.create(x,y,z)
+    try:
+        return Job.objects.create(id=payload["id"])
+    except ObjectDoesNotExist:
+        return None  # what to put here?
```
[the remark this answers is not in the document]
Comment (jaredwebber) [reply backend/job/query.py:17]: I think the plan is to have separate get and create methods - create takes the whole job payload while get just takes the job_id (and user_id for verification)

File: backend/job/models.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#         return {
+#             "id": self.id,
+#             "title": self.title,
+#             "date": self.date,
+#         }
+
+
+class Job(models.Model):
+    # column_id commented as causing tests to fail
+    # column_id = models.ForeignKey(KanbanColumn, on_delete=models.CASCADE)
```
Comment (jaredwebber) [inline backend/job/models.py:18]: This should be uncommented - It was only commented in the first place as I didn't want to troubleshoot the failing tests at midnight

File: backend/job/query.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LokiFrostGiant) [reply backend/job/query.py:17]: Confirming @jaredwebber 's analysis. Updated to create

File: backend/job/query.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        # error-free.
+        job.save()
+
+
+def get_job_by_id(in_user_id: int, job_id: int) -> Job:
+    job = Job.objects.get(id=job_id)
+    if job.user_id.id == in_user_id:
+        return job
+    else:
+        raise ObjectDoesNotExist("Job with that User does not exist")
```
[the remark this answers is not in the document]
Comment (LokiFrostGiant) [reply backend/job/query.py:46]: I will try this. I spent a bunch of time making the functions take in User, but this way will be cleaner if it works. Fingers crossed.

File: backend/job/models.py (hunk 2 above)
Comment (LokiFrostGiant) [reply backend/job/models.py:18]: leaving commented until the direct tests can pass with just User as a Foreign Key

File: backend/job/query.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (LokiFrostGiant) [reply backend/job/query.py]: made this work using your suggestion!

File: backend/job/query.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+def get_job_by_id(in_user: int, job_id: int) -> Job:
+    try:
+        return Job.objects.get(id=job_id, user__id=in_user)
+    except:
+        raise ObjectDoesNotExist("Job with that User does not exist")
+
+
+def get_minimum_jobs(in_user: int) -> QuerySet:
+    # return {id, column, position, company} for all user_id user's jobs
```
Comment (jaredwebber) [inline backend/job/query.py:59]: Could add `type` to this comment - or delete comment

File: backend/job/tests/test_views.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                                    content_type='application/json')
+
+        # Call the view function
+        response = views.update_job(request)
+
+        # Check the response
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.content.decode('utf-8'), json.dumps({}))
+
+        # Check that business.update_job was called
```
Comment (jaredwebber) [inline backend/job/tests/test_views.py:101]: This check isn't implemented

**Pull request merged**